### DISCUSSION TITLE
Cannot colour dagesh separately in Hebrew

### DISCUSSION THREAD

**Discussion by Drkazmer** [harfbuzz/harfbuzz#4454] [Q&A] [answered]
I have an issue, which is probably related to the rendering, namely that my commands to colour a _dagesh_ or a _shin/sin_ dot is ignored in LuaLaTeX. Please see my code example below:

```
\documentclass{minimal}
\usepackage{luacolor}
\usepackage[nil,bidi=basic]{babel}
\babelprovide[import=he,main]{hebrew}
\babelfont[hebrew]{rm}[Renderer=Harfbuzz,Script=Hebrew,Language=Default]{Times New Roman}

\begin{document}

\textcolor{blue}{ב}\textcolor{yellow}{ּ}\textcolor{red}{ָ}\textcolor{green}{֑}

\end{document}
```

See also post at StackExchange:

https://tex.stackexchange.com/q/698798/135685

**Comment by behdad** [comment, marked as answer]
This is probably because the dagesh forms a ligature with the base letter. If that is the case, unfortunately you cannot achieve what you want with this font.

**Reply by Drkazmer** [reply]
As I don't know its precise mechanics, hence whether that mattered either, but if I copy the letter with dagesh from the PDF and check its Unicode, it returns a value from the 05DX-05EX block and not the ligatures with dagesh in the FB3X-FB4X block.

**Reply by khaledhosny** [reply]
The font has a lookup under `ccmp` feature that turns this sequence into a ligature:
```
$ hb-shape -u 05D1,05BC "Times New Roman.ttf"
[uniFB31=0+932]
$ hb-shape -u 05D1,05BC "Times New Roman.ttf" --features="-ccmp"
[uni05BC=0+0|uni05D1=0+932]
```

**Reply by Udi-Fogiel** [reply]
@khaledhosny are there any risks in adding a callback to `pre_linebreak_filter` after harfbuzz to add the base letters over the ligature? This way we can color the base letter and the ligature in different colors. To my knowledge, ligatures involving hebrew nikud only affects the nikud positioning so the base letter should overlap the ligature  in a desired way. 

The only strange thing I encountered while doing that (see my answer to the stack exchange post) are nodes with character code outside of the Unicode span. I wasn't sure what is their meaning.

**Reply by khaledhosny** [reply]
I think you will get duplicate characters when extracting text from PDF (copying or searching) unless you wrap the glyphs in `/ActualText` spans. You will also be working with glyph IDs not characters at this point, and handling this is different depending on whether you are using HarfBuzz or the builtin luaotfload (`node`) shaper.

**Comment by Udi-Fogiel** [comment]
> I think you will get duplicate characters when extracting text from PDF (copying or searching) unless you wrap the glyphs in `/ActualText` spans.

That's a good point, I did not think about it. 

>You will also be working with glyph IDs not characters at this point, and handling this is different depending on whether you are using HarfBuzz or the builtin luaotfload (`node`) shaper.

I see. Do you to know where I can read more about how to correctly handle glyph IDs (with and without HarfBuzz)? Is the documentation of luaotfload has all the information, or should I read the HarfBuzz manuals as well?

**Reply by khaledhosny** [reply]
I don’t remember the details, it is been a long time, but I think luaotfload sets a property on glyph nodes after shaping with HarfBuzz that has the textual information. See https://github.com/latex3/luaotfload/blob/main/src/luaotfload-harf-plug.lua#L939 and search in this file how these node properties are used.

**Reply by Udi-Fogiel** [reply]
Thank for the reference!
